File: day02/c/part2.c
#include "helper.h"

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

const int GOAL = 19690720;


void debug(const int *numbers)
{
    for (int i = 0; i < arrlen(numbers); ++i) {
        printf("%d ", numbers[i]);
    }
    puts("");
}

int * process(const int *numbers)
{
    // start: make a copy
    int *data = NULL;
    arrsetlen(data, arrlen(numbers));
    for (int i = 0; i < arrlen(numbers); ++i) {
        data[i] = numbers[i];
    }
    // end

    int idx = 0;
    int opcode = data[idx];
    while (opcode != 99) {
        if (opcode == 1) {
            const int inp1 = data[data[idx + 1]];
            const int inp2 = data[data[idx + 2]];
            data[data[idx + 3]] = inp1 + inp2;
        }
        else if (opcode == 2) {
            const int inp1 = data[data[idx + 1]];
            const int inp2 = data[data[idx + 2]];
            data[data[idx + 3]] = inp1 * inp2;
        }
        idx += 4;
        opcode = data[idx];
    }

    return data;
}

int main()
{
    char* *lines = read_lines("input.txt");
    char* line = lines[0];
    bool terminate = false;

    int *numbers = extract_ints_from_string(line, ",");

    for (int i = 0; i <= 99; ++i) {
        for (int j = 0; j <= 99; ++j) {
            numbers[1] = i;
            numbers[2] = j;
            printf("# i = %d, j = %d\n", i, j);
            int *result = process(numbers);
            bool found = (result[0] == GOAL);
            arrfree(result);
            if (found) {
                printf("i = %d\n", i);
                printf("j = %d\n", j);
                printf("answer: %d\n", 100 * i + j);
                terminate = true;
                break;
            }
        }
        if (terminate) {
            break;
        }
    }
    puts("__END__");

    return 0;
}

File: day02/python/part2.py
#!/usr/bin/env python3

from typing import List

import helper

GOAL = 19690720


def process(numbers: List[int]) -> List[int]:
    data = numbers[:]
    idx = 0
    opcode = data[idx]
    while opcode != 99:
        if opcode == 1:
            inp1 = data[data[idx + 1]]
            inp2 = data[data[idx + 2]]
            data[data[idx + 3]] = inp1 + inp2
        elif opcode == 2:
            inp1 = data[data[idx + 1]]
            inp2 = data[data[idx + 2]]
            data[data[idx + 3]] = inp1 * inp2
        # endif
        idx += 4
        opcode = data[idx]
    # endwhile
    return data


def main() -> None:
    line = helper.read_lines("input.txt")[0]
    numbers = [int(n) for n in line.split(",")]
    terminate = False

    for i in range(0, 99+1):
        for j in range(0, 99+1):
            numbers[1] = i
            numbers[2] = j
            print(f"# i = {i}, j = {j}")
            result = process(numbers)
            if result[0] == GOAL:
                print("i =", i)
                print("j =", j)
                print("answer:", 100 * i + j)
                terminate = True
                break
            # endif
        # endfor
        if terminate:
            break
    # endfor
    print("__END__")

##############################################################################

if __name__ == "__main__":
    main()
